haskell/assets/gen-patch-win.py

```diff
@@ -0,0 +1,122 @@
+#!/usr/bin/env python
+
+import os
+import importlib
+import importlib.util
+import importlib.machinery
+import urllib.request
+import sys
+import hashlib
+import tarfile
+import tempfile
+from pathlib import PurePosixPath, Path
+from subprocess import Popen, PIPE, STDOUT, DEVNULL, check_call
+
+
+def calculate_sha256(file_path):
+    sha256_hash = hashlib.sha256()
+
+    with open(file_path, "rb") as file:
+        # Read the file in chunks to handle large files efficiently
+        chunk_size = 4096
+        chunk = file.read(chunk_size)
+        while len(chunk) > 0:
+            sha256_hash.update(chunk)
+            chunk = file.read(chunk_size)
+
+    return sha256_hash.hexdigest()
+
+def import_path(path):
+    module_name = os.path.basename(path).replace('-', '_')
+    spec = importlib.util.spec_from_loader(
+        module_name,
+        importlib.machinery.SourceFileLoader(module_name, path)
+    )
+    module = importlib.util.module_from_spec(spec)
+    spec.loader.exec_module(module)
+    sys.modules[module_name] = module
+    return module
+
+def need_to_download(file_path, sha256_hash):
+    if not os.path.exists(file_path):
+        return True
+    else:
+        return calculate_sha256(file_path).upper() != sha256_hash.upper()
+
+def download_file(url, file_path):
+    urllib.request.urlretrieve(url, file_path)
+
+def main():
+    if len(sys.argv) != 2:
+        print("usage: gen-patch-win.py GHC_VERSION", file=sys.stderr)
+        sys.exit(1)
+
+    VERSION = sys.argv[1]
+
+    # Download the ghc version for windows from the URL given in ghc_bindist_generated.bzl
+
+    SCRIPT_PATH = Path(__file__)
+    bindist = import_path(str(SCRIPT_PATH.parent.parent.joinpath('private/ghc_bindist_generated.bzl')))
+
+    ghc_url, hash = bindist.GHC_BINDIST[VERSION]['windows_amd64']
+
+    file_path = f'ghc-{VERSION}.windows_amd64.tar.xz'
+
+    if need_to_download(file_path, hash):
+        print("downloading...", ghc_url, file=sys.stderr)
+        download_file(ghc_url, file_path)
+
+        assert not need_to_download(file_path, hash), "file disappeared or hash mismatch"
+    else:
+        print("found existing file", file_path, file=sys.stderr)
+
+    print('extracting...', file=sys.stderr)
+
+    with tarfile.open(file_path, "r:xz") as tar:
+        temp_dir = tempfile.mkdtemp()
```

Comment:
Maybe this change? To remove the temporary directory at the end (or if something goes wrong).
```suggestion
    with tarfile.open(file_path, "r:xz") as tar, tempfile.TemporaryDirectory() as temp_dir:
```